opm/simulators/linalg/GraphColoring.hpp

```diff
@@ -218,5 +222,94 @@ reorderVerticesSpheres(const std::vector<int>& colors,
     }
     return indices;
 }
+
+/// \brief Specify coloring type.
+/// \details The coloring types have been implemented initially to parallelize DILU
+///          preconditioner and parallel sparse triangular solves.
+///          Symmetric coloring will create a dependency from row i to j if
+///          both element A_ij and A_ji exists.
+///          Lower coloring creates a dependency from row i to j (where i < j) if
+///          A_ij is nonzero.
+///          Upper coloring creates a dependecy from row i to j (where i > j) if A_ij is nonzero.
+enum class ColoringType { SYMMETRIC, LOWER, UPPER };
+
+/// This coloring algorithm interprets the sparsity structure of a matrix as a graph. Each row is given a color
+/// or level where all the rows in the same level only have dependencies from lower levels. The level computation
+/// is done with dynamic programming, and to improve caching the rows in the same level stay in matrix order.
+/// \brief Given a matrix and dependecy type, returns a SparseTable grouping the rows by which
+///        can be executed in parallel without breaking dependencies
+/// \param matrix A dune sparse matrix
+/// \param coloringType The coloringtype determines what constitutes a dependency,
+///        see ColoringType definition above
+/// \return SparseTable with rows of the matrix grouped into least number of groups
+///         while dependencies only come from groups with lower index
+template <class M>
+Opm::SparseTable<std::size_t>
+getMatrixRowColoring(const M& matrix, ColoringType coloringType)
+{
+    OPM_TIMEBLOCK(createMatrix);
+
+    std::vector<std::size_t> color(matrix.N(), 0);
+    std::vector<std::size_t> rowIndices(matrix.N(), 0);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+    for (std::size_t i = 0; i < matrix.N(); i++) {
```

Comment:
no point parallelizing this loop i don't think, and thus
 ```
std::iota(rowIndices.begin(), rowIndices.end(), 0);
```
is simpler and more expressive.

Reply:
Why is there no point, will this not run faster with multiple threads?

Reply:
maybe it would run slightly faster if threads end up on processors with separate memory lanes but it is really nothing you'd be able to measure. there is basically no computation here, only writing to memory.